### PR TITLE
Init: Find which presubmits have to be executed.

### PR DESCRIPTION
**Pull request by droslean**
Compares the `presubmits` from the decorated repository and the master files.
Finds which of them shall be executed and creates a mapping of the repo to `presubmits` to execute.

### CODE REVIEW

File: cmd/pj-rehearse/main.go
Comment (petr-muller) [inline cmd/pj-rehearse/main.go]: I am not that happy with the `decorationPath` name, as it does not really describe what it is. I think we should accept symmetrical options for master and under-test configs. Either accept paths to full `openshift/release` repo clones (one for master, one for PR) and compute the `jobConfig` paths by knowing the layout of `openshift/release`, or accept two sets of options (again, one for master, one for pr) pointing to needed directories directly.

File: pkg/diffs/diffs.go
Comment (petr-muller) [inline pkg/diffs/diffs.go]: I think do not assume what the Presubmits are for. They are Presubmits affected by a proposed change, what they are used for is not this code's business. The ...ToExecute suffix increases the code coupling unnecessarily.

File: pkg/diffs/diffs.go
Comment (petr-muller) [inline pkg/diffs/diffs.go]: I think this will crash on newly added jobs. Newly added job's jobName won't have an appropriate master presubmit item, and we will attempt to access such item's Spec.

Also, are we really sure we do not want to rehearse jobs changed in non-Spec fields? Are all these fields that harmless?

@droslean 's comment in the original #50:
> We will have to be more specific in which fields we need to execute the job. Here I only check the .spec.

What do you mean about being more specific?

File: pkg/diffs/diffs.go
Comment (petr-muller) [inline pkg/diffs/diffs.go]: Do we need to pollute the names with type information (the `Map` suffix)? It's rarely useful elsewhere but maybe I'm missing some Go folk knowledge.

File: pkg/diffs/diffs.go
Comment (petr-muller) [inline pkg/diffs/diffs.go]: The path contatenation should be done with `filepath.Join` istead of `fmt.Sprintf`.

File: pkg/diffs/diffs.go
Comment (petr-muller) [inline pkg/diffs/diffs.go:39]: Maybe a small comment explaining why this function needs to exist (along the lines of the original thread here https://github.com/openshift/ci-operator-prowgen/pull/50#discussion_r245327669) whould be helpful for the future readers.

**Review by petr-muller — COMMENTED** [review]
I like this version. It's nice how it got really straightforward and simple. Let's iron out the code comments (inline, mostly taken from the original #50 PR) and let's get it merged so that we finally have a base to iterate on (I can then rebase my PR on this easily). I would even say let's merge it without tests and proivde them in a followup PR.

File: pkg/diffs/diffs.go
Comment (droslean) [reply pkg/diffs/diffs.go]: It won't crash, I also tested it with newly added repo as well.  What other filed we should consider worth to rehearse except the .spec?

File: pkg/diffs/diffs.go
Comment (droslean) [reply pkg/diffs/diffs.go]: with `make` I make sure that this map won't be nil. In that case, it will create an empty map.

Comment (droslean) [reply pkg/diffs/diffs.go]: Also, the reason why it won't crash mentioned in [#comment](https://github.com/openshift/ci-operator-prowgen/pull/52#discussion_r245669029)

File: cmd/pj-rehearse/main.go
Comment (bbguimaraes) [inline cmd/pj-rehearse/main.go:1]: It will be easier to integrate this if you use `cmd/rehearse/main.go` from Petr's PR.

File: cmd/pj-rehearse/main.go
Comment (bbguimaraes) [inline cmd/pj-rehearse/main.go]: Missing a new-line character here (mentioning just in case it gets accidentally transplanted to the actual `main.go` file).

File: pkg/diffs/diffs.go
Comment (bbguimaraes) [reply pkg/diffs/diffs.go]: Following up from Petr's earlier suggestion, this variable should be named differently. Personally, I would choose `ret`: short and instantly conveys its purpose.

File: pkg/diffs/diffs.go
Comment (bbguimaraes) [inline pkg/diffs/diffs.go]: No need for these declarations if the variables are not used until L41.

File: pkg/diffs/diffs.go
Comment (bbguimaraes) [inline pkg/diffs/diffs.go]: `prowPRConfig.JobConfig.Presubmits` can be used to drive the loop, there is no need to create a `map` for it.

File: pkg/diffs/diffs.go
Comment (bbguimaraes) [reply pkg/diffs/diffs.go]: Accessing non-existing keys will yield an uninitialized map followed by an uninitialized `ProwJobSpec`, which should be fine if `DeepEqual` can handle the later.

It may be faster to explicitly check for their existence, but the tool seemed fast enough in my tests.

File: pkg/diffs/diffs.go
Comment (bbguimaraes) [reply pkg/diffs/diffs.go:39]: The function can be inlined if we remove the unnecessary call above.

**Review by bbguimaraes — CHANGES_REQUESTED** [review]
A couple of comments, mostly on implementation details.

@droslean, @petr-muller: how do we want to combine these PRs? I made mine a super-set of Petr's so they can be merged trivially via a fast-forward merge (before or after the merge to master).

**Comment by petr-muller** [comment]
> how do we want to combine these PRs? I made mine a super-set of Petr's so they can be merged trivially via a fast-forward merge (before or after the merge to master).

I don't mind rebasing my PR (which up to now served more like an experiment, showcase vessel and not a real PR) on top of this. So I vote to resolve the little things here and merge it, IMO it's the easiest way.

File: pkg/diffs/diffs.go
Comment (petr-muller) [reply pkg/diffs/diffs.go]: Sorry, I spent too much time in my life using languages where uninitialized stuff bites you in the butt.

File: pkg/diffs/diffs.go
Comment (droslean) [reply pkg/diffs/diffs.go]: I can use it to iterate, but I can't use it to compare it with the master's `prowPRConfig.JobConfig.Presubmits`.

File: pkg/diffs/diffs.go
Comment (stevekuznetsov) [inline pkg/diffs/diffs.go]: What's the utility of this struct?

File: pkg/diffs/diffs.go
Comment (stevekuznetsov) [inline pkg/diffs/diffs.go]: If they're exported fields, why bother with a constructor?

File: pkg/diffs/diffs.go
Comment (stevekuznetsov) [inline pkg/diffs/diffs.go:19]: Err - why are these hard-coded?

File: pkg/diffs/diffs.go
Comment (stevekuznetsov) [inline pkg/diffs/diffs.go]: This is somehow expecting the PR to be checked out in a different location from the current working directory. In what set-up is this true? Will this be true in a PR job? Why have we chosen this?

File: pkg/diffs/diffs.go
Comment (stevekuznetsov) [reply pkg/diffs/diffs.go]: There are dozens of fields outside of spec. Why not diff on them all? If I change an `agent: kubernetes` job to `agent: jenkins` I would expect this rehearsal to catch it.

**Review by stevekuznetsov — COMMENTED** [review]
I think this portion of the rehearsal should start from the working directory of a job - PR is merged into the target branch. Can we use `git` and other tools to determine what the config state was before and after the proposed changes?

File: pkg/diffs/diffs.go
Comment (droslean) [reply pkg/diffs/diffs.go:19]: We assume that those files are in the parent directory because of the decoration.

**Comment by droslean** [comment]
@stevekuznetsov Answering to your [comment](https://github.com/openshift/ci-operator-prowgen/pull/52#discussion_r245847063): If we need to check all the fields of the job then we can't use `equality.Semantic.DeepEqual` because the struct contains unexporting fields. I think also that its better to pick the specific fields that we want to compare, to avoid reruning the jobs for changes in fields like `rerun_command` etc...

File: pkg/diffs/diffs.go
Comment (bbguimaraes) [reply pkg/diffs/diffs.go]:
```go
masterJobs := getJobsByRepoAndName(prowMasterConfig.JobConfig.Presubmits)
for repo, jobs := range prowPRConfig.JobConfig.Presubmits {
    for _, job := range jobs {
        if !equality.Semantic.DeepEqual(masterJobs[repo][job.Name].Spec, job.Spec) {
            preSubmitsToExecuteMap[repo] = append(preSubmitsToExecuteMap[repo], job)
        }
    }
}
```

Comment (droslean) [reply pkg/diffs/diffs.go]: good catch. updated

**Comment by petr-muller** [comment]
We discussed this on Slack, here's the summary: the only remaining big question was the availability of both PR and master configs to the Prow job running the rehearse. PR version is available via decoration and master version is available in the cluster and it can be mounted to the Prowjob. We agreed we will merge this PR as a base for further development and set up full final `openshift/release` presubmit to run it (https://github.com/openshift/release/pull/2520, https://github.com/openshift/release/pull/2522). Right now that presubmit will only list the jobs that will be rehearsed in the future, which should be totally safe. We can incrementally develop it and this early deployment can keep us honest about not merging broken stuff :)

/lgtm
/hold for https://github.com/openshift/release/pull/2520 to merge so it can test this PR.

File: pkg/diffs/diffs.go
Comment (stevekuznetsov) [reply pkg/diffs/diffs.go:19]: Let's not do this.

**Pull request merged**
